client/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -0,0 +1,476 @@
+            if (hosts == null || hosts.length == 0) {
+                throw new IllegalArgumentException("no hosts provided");
+            }
+            this.hosts = hosts;
+            return this;
+        }
+
+        /**
+         * Sets the default request headers, to be used when creating the default http client instance.
+         * In case the http client is set through {@link #setHttpClient(CloseableHttpClient)}, the default headers need to be
+         * set to it externally during http client construction.
+         */
+        public Builder setDefaultHeaders(Header[] defaultHeaders) {
+            Objects.requireNonNull(defaultHeaders, "default headers must not be null");
+            for (Header defaultHeader : defaultHeaders) {
+                Objects.requireNonNull(defaultHeader, "default header must not be null");
+            }
+            this.defaultHeaders = defaultHeaders;
+            return this;
+        }
+
+        /**
+         * Creates a new {@link RestClient} based on the provided configuration.
+         */
+        public RestClient build() {
+            if (httpClient == null) {
+                httpClient = createDefaultHttpClient(null);
+            }
+            if (hosts == null || hosts.length == 0) {
+                throw new IllegalArgumentException("no hosts provided");
+            }
+            return new RestClient(httpClient, maxRetryTimeout, defaultHeaders, hosts);
+        }
+
+        /**
+         * Creates a {@link CloseableHttpClient} with default settings. Used when the http client instance is not provided.
+         *
+         * @see CloseableHttpClient
+         */
+        public static CloseableHttpClient createDefaultHttpClient(Registry<ConnectionSocketFactory> socketFactoryRegistry) {
+            PoolingHttpClientConnectionManager connectionManager;
+            if (socketFactoryRegistry == null) {
+                connectionManager = new PoolingHttpClientConnectionManager();
+            } else {
+                connectionManager = new PoolingHttpClientConnectionManager(socketFactoryRegistry);
+            }
+            //default settings may be too constraining
+            connectionManager.setDefaultMaxPerRoute(10);
+            connectionManager.setMaxTotal(30);
+
+            //default timeouts are all infinite
+            RequestConfig requestConfig = RequestConfig.custom().setConnectTimeout(DEFAULT_CONNECT_TIMEOUT)
+                    .setSocketTimeout(DEFAULT_SOCKET_TIMEOUT)
+                    .setConnectionRequestTimeout(DEFAULT_CONNECTION_REQUEST_TIMEOUT).build();
+            return HttpClientBuilder.create().setConnectionManager(connectionManager).setDefaultRequestConfig(requestConfig).build();
+        }
+    }
+
+    /**
+     * Listener that allows to be notified whenever a failure happens. Useful when sniffing is enabled, so that we can sniff on failure.
+     * The default implementation is a no-op.
+     */
+    public static class FailureListener {
```

Comment:
I tend to agree, I just didn't want to have an interface and a default empty impl. default methods would help :(